src/structure.js

```diff
@@ -53,10 +55,32 @@ function analyzeType(value) {
   };
 }
 
-function isa(Child, Ancestor) {
+export function isa(Child, Ancestor) {
   return Child === Ancestor || Child.prototype instanceof Ancestor;
 }
 
+export function collapseState(tree, value) {
+  return new $(tree)
+      .flatMap(node => {
+        if (node.isSimple) {
+          if (isa(node.Type, types.Array)) {
+            return analyzeType(value)(new Node(types.Array, node.path))
+          }
+          if (isa(node.Type, types.Object)) {
+            return analyzeType(value)(new Node(types.Object, node.path));
+          }
+        }
+        return analyzeType(value)(node);
+      })
+    .map(node => {
+      if (node.isSimple) {
+        return node.valueAt(value) || new node.Type(node.valueAt(value)).valueOf();
```

Comment:
Can we not push this logic into `node.stateAt`? so that the map is just `map(node => node.stateAt(value)`?